simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/support/ChangeAvailabilityFuture.java

```diff
@@ -0,0 +1,68 @@
+package com.evbox.everon.ocpp.simulator.station.handlers.ocpp.support;
+
+import com.evbox.everon.ocpp.simulator.station.evse.Evse;
+import com.evbox.everon.ocpp.simulator.station.evse.EvseState;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class ChangeAvailabilityFuture {
+
+    private final ExecutorService threadPool;
+
+    /**
+     * Create an instance with the threadPool sized to what {@code Runtime.getRuntime().availableProcessors()} return
+     */
+    public ChangeAvailabilityFuture() {
+        this.threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+    }
+
+    /**
+     * Create an instance with the given threadPool.
+     *
+     * @param threadPool
+     */
+    public ChangeAvailabilityFuture(ExecutorService threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Tries to set a new state withing a given time frame. It sets the new state if evse has no ongoing transaction.
+     *
+     * @param evse                         {@link Evse}
+     * @param newState                     a new evse state
+     * @param attempts                     number of attempts to set a new state
+     * @param delayBetweenAttemptsInMillis delay between attempts in milliseconds
+     */
+    public void runAsync(Evse evse, EvseState newState, int attempts, long delayBetweenAttemptsInMillis) {
+
+        CompletableFuture.runAsync(() -> {
+            int count = 0;
+
+            while (count++ < attempts) {
```

Comment:
why do we limit the number of attempts?
And instead of waiting I'd think about event listeners and change Evse state when onTransactionFinished event is triggered

Reply:
Give up strategy. Dont block thread forever if exit-condition does not return true.

Who triggers onTransactionFinished event? As of now, transaction status might change either from server or from user. 

Did I understand you correctly, that you are proposing to have a dedicated memory-based internal mechanism of communication? For example: user stops transaction -> we send an event to memory-queue -> consumer receives an event -> changes status of the transaction.

Reply:
I removed this class.

Reply:
I proposed something more simple. You have a single place where transaction stops, even though it may be initiated by different reasons and from this single place you either send a message to interesting parties (I mean not OCPP message but sort of event notification) or simply apply these scheduled things. 

Implementation details it's another question but extra thread for this is definitely an overkill.

Reply:
I changed that logic.